Import viz code as a library

jameshadfield opened the pull request:
This commit moves to a design pattern with:
* pathogen-agnostic reusable visualisation components
* A page which renders SARS-CoV-2 specific visualisations using the above library

This commit creates the latter in this repo, with the former packed and vendored here as a tarball. This should be considered temporary, with this page moving to the nextstrain.org repo once some technical hurdles have been overcome there.

This webpage is currently only being used to produce static figures via a GitHub action which are then sourced by a static page on nextstrain.org via https://github.com/nextstrain/nextstrain.org/pull/660. To reduce any confusion, the abstracts of each figure have been removed from this repo to make it clear that the abstracts on nextstrain.org are the authoritative versions.

Viz library packed from https://github.com/nextstrain/forecasts-viz/tree/05ac579a885d727bbaac4d9dbc63198b0d155d38

joverlee521 commented:
I didn't dig into the code changes, but just tested the app and the `static-images.js` script. 

The one thing I noticed is that the panel figures created from this PR all include their own legend. 
If this is intentional, then I can remove the separate legends in https://github.com/nextstrain/nextstrain.org/pull/660

jameshadfield commented:
@joverlee521 I've updated the `static-images.js` script to create figures at various screen resolutions and added an example HTML page using `srcset`s to choose the appropriate image to load. If we are happy with this we can then update https://github.com/nextstrain/nextstrain.org/pull/660 accordingly

joverlee521 commented on viz/example-static-images.html:
```diff
@@ -0,0 +1,44 @@
+      <div class="center">
+        <!-- the widths used in the img were taken from the script's STDOUT -->
+        <img
+          srcset="./figures/frequenciesPanel_small.png 504w, ./figures/frequenciesPanel_medium.png 784w, ./figures/frequenciesPanel_large.png 1084w, ./figures/frequenciesPanel_xlarge.png 1334w"
+          sizes="(max-width: 999px) 504px, (max-width: 1299px) 784px, (max-width: 1549px) 1084px, 1334px,"
+          src="./figures/frequenciesPanel_small.png"
```
Hmm, this isn't working for me...

The image changes from `xlarge` to `large`, but never changes to the `medium` or `small` versions even when I'm viewing with the FireFox "Responsive Design Mode".

joverlee521 replied:
OH, if I change to `DPR: 1` then the resizing works as expected.

joverlee521 replied:
Hmm, maybe it's more appropriate to use the [picture element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture) here.

joverlee521 replied:
See update in 22879342b9874ce0edc3e00f2148f904f53d1538

jameshadfield replied:
Nice! Easier to read as well :)

joverlee521 commented:
(CI failed due to expired cert)

jameshadfield commented on viz/scripts/static-images.js:
```diff
@@ -48,32 +60,22 @@ async function captureScreenshot(dir) {
+      console.log(`Finding & saving panels at size ${resolution.name} (page width of ${resolution.width}px)`);
+      for (const id of elementsToScreenshot) {
+        const panel = await page.$(`#${id}`);
+        const figureWidth = Math.floor((await panel.boundingBox()).width);
+        const fname = `${dir}/${id}_${resolution.name}.png`
```
The values used in `srcset` are more of a guide, as I understand it. But you're right that we do want to catch things as they drift here. How about I use the actual figure-widths in the filename and `assert` that they are as we expect in the image-generation script? If this script fails will we notice it in our CI?

joverlee521 replied:
Oh, we actually don't even need to include widths in `srcset` if we use the `media` tag instead of `sizes`. 

Hmm, I'm okay with keeping the general small/medium/large/xlarge labels if we don't care about the specific widths and only want them to fit the general min-width cutoffs.

jameshadfield replied:
Let's just go with what we've got - this entire workflow is hopefully short-term after all!

joverlee521 reviewed (approved):
This looks good to merge!

I ran the [Static Model Viz workflow](https://github.com/nextstrain/forecasts-ncov/actions/runs/4546524808/jobs/8015257157) with this branch and the trial figures are available at https://nextstrain-data.s3.amazonaws.com/files/workflows/forecasts-ncov/gisaid/trial/viz-library-mk2/nextstrain_clades/global/ 

I'll update https://github.com/nextstrain/nextstrain.org/pull/660 to use the different figures and remove the additional legends.
